tsdb/tombstones/tombstones.go

```diff
@@ -252,6 +252,32 @@ func (t *MemTombstones) Get(ref uint64) (Intervals, error) {
 	return t.intvlGroups[ref], nil
 }
 
+func (t *MemTombstones) DeleteTombstones(refs map[uint64]struct{}) {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+	for ref := range refs {
+		delete(t.intvlGroups, ref)
+	}
+}
+
+func (t *MemTombstones) TruncateBefore(beforeT int64) {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+	for ref, ivs := range t.intvlGroups {
+		i := len(ivs) - 1
+		for ; i >= 0; i-- {
+			if beforeT > ivs[i].Maxt {
+				break
+			}
+		}
+		if len(ivs[i+1:]) == 0 {
+			delete(t.intvlGroups, ref)
+		} else {
+			t.intvlGroups[ref] = ivs[i+1:]
```

Comment:
Note to myself: check if this cleans up memory for `ivs[:i+1]`, if not, create a new slice and copy over the items instead.

Reply:
>  Since the slice references the original array, as long as the slice is kept around the garbage collector can’t release the array; the few useful bytes of the file keep the entire contents in memory.

[ref](https://go.dev/blog/slices-intro)
I think you are right, it doesn't reclaim memory in my experiments locally too. 

pushing a fix for this 👍